sonar-plugin-api/src/main/java/org/sonar/api/server/rule/internal/DefaultNewRule.java

```diff
@@ -336,6 +341,30 @@ void validate() {
     if (isEmpty(htmlDescription) && isEmpty(markdownDescription)) {
       throw new IllegalStateException(format("One of HTML description or Markdown description must be defined for rule %s", this));
     }
+    validateSameContextKeysExistsForAllContextualizedSections(ruleDescriptionSections);
+  }
+
+  private static void validateSameContextKeysExistsForAllContextualizedSections(List<RuleDescriptionSection> ruleDescriptionSections) {
+    Map<String, Set<String>> sectionKeyToContextKeys = ruleDescriptionSections.stream()
+      .filter(section -> section.getContext().isPresent())
+      .collect(groupingBy(RuleDescriptionSection::getKey, mapping(s -> s.getContext().get().getKey(), Collectors.toSet())));
+
+    assertAllSectionsContainsSameContextKeys(sectionKeyToContextKeys);
+  }
+
+  private static void assertAllSectionsContainsSameContextKeys(Map<String, Set<String>> sectionKeyToContextKeys) {
+    if (sectionKeyToContextKeys.isEmpty()) {
+      return;
+    }
+
+    Map.Entry<String, Set<String>> referenceSectionToContexts = sectionKeyToContextKeys.entrySet().iterator().next();
+    String referenceSectionKey = referenceSectionToContexts.getKey();
+    Set<String> referenceContexts = referenceSectionToContexts.getValue();
+
+    sectionKeyToContextKeys.forEach((sectionKey, contextKeys) ->
```

Comment:
Also nitpick: if you keep a reference on the iterator, you could use `forEachRemaining` to avoid comparing the first element with itself

Reply:
I gave it a try and it would force to iterate on the entries of the map, instead of having the key value already split in 2 lambda parameters... so for readability I'd stick with this one.